File: knotlink-sdk-qt/tests/test_request_response_integration.cpp
/*
 * KnotLink SDK - Qt C++ Integration Test
 * Request-Response pattern: Responser + Querier
 */

#include <QCoreApplication>
#include <QThread>
#include <QTimer>
#include <QDebug>
#include "../knotlink/opensocketresponser.h"
#include "../knotlink/opensocketquerier.h"

static bool shutdown = false;

void run_responser() {
    OpenSocketResponser res("app.knotlinksdk.test", "test_socket");
    QObject::connect(&res, &OpenSocketResponser::receivedData,
        [](const QString &data, QString) {
            qDebug() << "[Responser] Received:" << data;
        });
    qDebug() << "[Responser] Running...";
    while (!shutdown) {
        QThread::msleep(1000);
    }
}

void run_querier() {
    QThread::msleep(1000);
    OpenSocketQuerier q("app.knotlinksdk.test", "test_socket");
    QString result = q.query_l("Hello, Responser!");
    qDebug() << "[Querier] Response:" << result;
    shutdown = true;
}

int main(int argc, char *argv[]) {
    QCoreApplication app(argc, argv);

    QThread* t = QThread::create(run_responser);
    t->start();

    QTimer::singleShot(100, run_querier);
    QTimer::singleShot(5000, &app, &QCoreApplication::quit);

    app.exec();
    t->wait();
    return 0;
}

File: knotlink-sdk-python/tests/integration/test_request_response_integration.py
#!/usr/bin/env python3
# test_request_response_integration.py
import time
import threading
from knotlink import OpenSocketResponser, OpenSocketQuerier

_shutdown = threading.Event()

def run_responser():
    def on_request(data):
        print(f"[Responser] Received: {data}")
        return f"Echo: {data}"
    r = OpenSocketResponser("app.knotlinksdk.test", "test_socket")
    r.set_RecvFunc(on_request)
    print("[Responser] Running...")
    try:
        while not _shutdown.wait(timeout=1):
            pass
    finally:
        r.disconnect()
        print("[Responser] Disconnected.")

def run_querier():
    _shutdown.wait(timeout=1)  # 等待 responser 启动
    q = OpenSocketQuerier("app.knotlinksdk.test", "test_socket")
    try:
        result = q.query("Hello, Responser!")
        print(f"[Querier] Response: {result}")
    except Exception as e:
        print(f"[Querier] Error: {e}")
    finally:
        q.disconnect()
        print("[Querier] Disconnected.")
        _shutdown.set()  # 通知 responser 退出

if __name__ == "__main__":
    t = threading.Thread(target=run_responser, daemon=True)
    t.start()
    run_querier()
    t.join(timeout=5)